app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -145,6 +155,76 @@ async function addEthereumChainHandler(
     );
   }
 
+  const switchChainWithPermissions = async ({
+    networkConfigurationId,
+    approvalFlowId,
+  }) => {
+    let permissionedChainIds;
+    try {
+      ({ value: permissionedChainIds } = getCaveat(
+        origin,
+        PermissionNames.wallet_switchEthereumChain,
+        CaveatTypes.restrictNetworkSwitching,
+      ));
+    } catch (e) {
+      // throws if the origin does not have any switchEthereumChain permissions yet
+      if (e instanceof PermissionDoesNotExistError) {
+        // suppress expected error in case that the domain does not have a
+        // wallet_switchEthereumChain permission set yet
+      } else {
+        throw e;
+      }
+    }
+
+    if (
+      permissionedChainIds === undefined ||
+      !permissionedChainIds.includes(_chainId)
+    ) {
+      try {
+        // TODO replace with caveat merging once merged
+        // rather than passing already permissionedChains here
+        await requestSwitchNetworkPermission([
+          ...(permissionedChainIds ?? []),
+          chainId,
+        ]);
+      } catch (err) {
+        res.error = err;
+        return end();
+      }
+    }
+
+    const networkClientId = findNetworkClientIdByChainId(chainId);
+
+    try {
+      await setActiveNetwork(networkConfigurationId ?? networkClientId);
+      if (hasPermissions(req.origin)) {
+        setNetworkClientIdForDomain(
+          req.origin,
+          networkConfigurationId ?? networkClientId,
+        );
+      }
```

Comment:
What does it mean to `setActiveNetwork` but not `setNetworkClientIdForDomain`?

Also, `hasPermissions` returns true if the subject has _any_ permissions, including `eth_accounts`. Is that what we want here?

Reply:
> What does it mean to setActiveNetwork but not setNetworkClientIdForDomain?

It means to set the globally selected network without setting the per dapp selected network state. Until we migrate the global network state to the SelectedNetworkController we have this ugly pattern

> Also, hasPermissions returns true if the subject has any permissions, including eth_accounts. Is that what we want here?

Good catch, but its actually the other way around, we only want it to be concerned with the `eth_accounts` permission in this case. We're checking the `permittedChains` permission above.

Reply:
Ok I've narrowed this to only query for the `eth_accounts` permission: https://github.com/MetaMask/metamask-extension/pull/24415/commits/c67f99d3992b89df200aac213ed9a74afd74569c

Reply:
Ah, got it. What's the motivation for "if the origin has the `eth_accounts` permission we set per dapp network selection state"?

Reply:
The idea is that we **don't** want to add this state unless there is `eth_account` permission since the state would then balloon to include every site the user visited. We've removed the logic for that to happen as it previously did in `setupProviderEngine`, but `setNetworkClientIdForDomain` will still set this state and create a new proxy for the origin if called with a new origin so we guard with a [check that the origin has permissions](https://github.com/MetaMask/core/blob/0a6bb9e029357964e1e3a1fb1261c9f89aed72ac/packages/selected-network-controller/src/SelectedNetworkController.ts#L306) and throw an error if it doesn't to alert devs against misuse